File: Packages/web-sdk/src/noFillTracker.ts
/**
 * NoFillTracker - Tracks no-fill events for analytics and pattern detection
 *
 * Features:
 * - Event recording with rich metadata
 * - Hourly and daily aggregation
 * - Pattern detection (consecutive failures, elevated rates)
 * - Statistics and breakdown by source, placement, reason
 */

/**
 * Reasons for ad no-fill
 */
export enum NoFillReason {
  TIMEOUT = 'timeout',
  NO_INVENTORY = 'no_inventory',
  NETWORK_ERROR = 'network_error',
  POLICY_VIOLATION = 'policy_violation',
  FREQUENCY_CAP = 'frequency_cap',
  GEOGRAPHIC_RESTRICTION = 'geographic_restriction',
  BUDGET_EXHAUSTED = 'budget_exhausted',
  MALFORMED_RESPONSE = 'malformed_response',
  SERVER_ERROR = 'server_error',
  UNKNOWN = 'unknown',
}

/**
 * A single no-fill event
 */
export interface NoFillEvent {
  sourceId: string;
  placementId: string;
  reason: NoFillReason;
  timestamp: number;
  latencyMs: number;
  metadata: Record<string, string>;
}

/**
 * Statistics for no-fill events
 */
export interface NoFillStats {
  totalNoFills: number;
  noFillRate: number; // Per minute
  averageLatencyMs: number;
  topReasons: Map<NoFillReason, number>;
  topSources: Map<string, number>;
  topPlacements: Map<string, number>;
  hourlyBreakdown: Map<number, number>;
  dailyBreakdown: Map<number, number>;
}

/**
 * Pattern type
 */
export enum PatternType {
  ELEVATED_RATE = 'elevated_rate',
  SOURCE_SPECIFIC = 'source_specific',
  PLACEMENT_SPECIFIC = 'placement_specific',
  CONSECUTIVE_FAILURES = 'consecutive_failures',
}

/**
 * Pattern severity
 */
export enum PatternSeverity {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical',
}

/**
 * Detected no-fill pattern
 */
export interface NoFillPattern {
  type: PatternType;
  severity: PatternSeverity;
  description: string;
  detectedAt: number;
  affectedSourceId?: string;
  affectedPlacementId?: string;
}

/**
 * Configuration for NoFillTracker
 */
export interface NoFillTrackerConfig {
  maxEventsRetained: number;
  maxRetentionHours: number;
  elevatedRateThreshold: number;
  patternDetectionEnabled: boolean;
  consecutiveFailureThreshold: number;
}

/**
 * Default configuration
 */
export const DEFAULT_CONFIG: NoFillTrackerConfig = {
  maxEventsRetained: 10000,
  maxRetentionHours: 24,
  elevatedRateThreshold: 0.5,
  patternDetectionEnabled: true,
  consecutiveFailureThreshold: 5,
};

/**
 * Pattern listener type
 */
export type NoFillPatternListener = (pattern: NoFillPattern) => void;

/**
 * NoFillTracker class for tracking and analyzing no-fill events
 */
export class NoFillTracker {
  private static instance: NoFillTracker | null = null;

  private config: NoFillTrackerConfig;
  private events: NoFillEvent[] = [];
  private totalNoFills = 0;
  private totalLatencyMs = 0;

  private hourlyNoFills: Map<number, number> = new Map();
  private dailyNoFills: Map<number, number> = new Map();
  private noFillsBySource: Map<string, number> = new Map();
  private noFillsByPlacement: Map<string, number> = new Map();
  private noFillsByReason: Map<NoFillReason, number> = new Map();
  private consecutiveNoFillsBySource: Map<string, number> = new Map();
  private detectedPatterns: NoFillPattern[] = [];
  private patternListeners: NoFillPatternListener[] = [];

  constructor(config: Partial<NoFillTrackerConfig> = {}) {
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  /**
   * Get singleton instance
   */
  static getInstance(): NoFillTracker {
    if (!NoFillTracker.instance) {
      NoFillTracker.instance = new NoFillTracker();
    }
    return NoFillTracker.instance;
  }

  /**
   * Reset singleton instance (for testing)
   */
  static resetInstance(): void {
    NoFillTracker.instance = null;
  }

  /**
   * Record a no-fill event
   */
  recordNoFill(
    sourceId: string,
    placementId: string,
    reason: NoFillReason,
    latencyMs: number = 0,
    metadata: Record<string, string> = {}
  ): void {
    const event: NoFillEvent = {
      sourceId,
      placementId,
      reason,
      timestamp: Date.now(),
      latencyMs,
      metadata,
    };

    this.events.push(event);
    this.totalNoFills++;
    this.totalLatencyMs += latencyMs;

    // Update source counter
    this.noFillsBySource.set(
      sourceId,
      (this.noFillsBySource.get(sourceId) || 0) + 1
    );

    // Update placement counter
    this.noFillsByPlacement.set(
      placementId,
      (this.noFillsByPlacement.get(placementId) || 0) + 1
    );

    // Update reason counter
    this.noFillsByReason.set(
      reason,
      (this.noFillsByReason.get(reason) || 0) + 1
    );

    // Update hourly breakdown
    const date = new Date(event.timestamp);
    const hour = date.getHours();
    this.hourlyNoFills.set(hour, (this.hourlyNoFills.get(hour) || 0) + 1);

    // Update daily breakdown
    const dayOfWeek = date.getDay();
    this.dailyNoFills.set(
      dayOfWeek,
      (this.dailyNoFills.get(dayOfWeek) || 0) + 1
    );

    // Update consecutive failures
    this.consecutiveNoFillsBySource.set(
      sourceId,
      (this.consecutiveNoFillsBySource.get(sourceId) || 0) + 1
    );

    // Cleanup old events
    this.cleanupIfNeeded();

    // Detect patterns
    if (this.config.patternDetectionEnabled) {
      this.detectPatternsForEvent(event);
    }
  }

  /**
   * Record a successful fill (resets consecutive failure counter)
   */
  recordFill(sourceId: string): void {
    this.consecutiveNoFillsBySource.set(sourceId, 0);
  }

  /**
   * Get current statistics
   */
  getStats(): NoFillStats {
    const avgLatency =
      this.totalNoFills > 0 ? this.totalLatencyMs / this.totalNoFills : 0;

    // Calculate rate per minute (based on last hour)
    const oneHourAgo = Date.now() - 3600000;
    const recentEvents = this.events.filter((e) => e.timestamp >= oneHourAgo);
    const ratePerMinute = recentEvents.length / 60;

    // Get top reasons (sorted by count)
    const topReasons = new Map(
      [...this.noFillsByReason.entries()]
        .sort((a, b) => b[1] - a[1])
        .slice(0, 5)
    );

    // Get top sources
    const topSources = new Map(
      [...this.noFillsBySource.entries()]
        .sort((a, b) => b[1] - a[1])
        .slice(0, 5)
    );

    // Get top placements
    const topPlacements = new Map(
      [...this.noFillsByPlacement.entries()]
        .sort((a, b) => b[1] - a[1])
        .slice(0, 5)
    );

    return {
      totalNoFills: this.totalNoFills,
      noFillRate: ratePerMinute,
      averageLatencyMs: avgLatency,
      topReasons,
      topSources,
      topPlacements,
      hourlyBreakdown: new Map(this.hourlyNoFills),
      dailyBreakdown: new Map(this.dailyNoFills),
    };
  }

  /**
   * Get hourly breakdown
   */
  getHourlyBreakdown(): Map<number, number> {
    return new Map(this.hourlyNoFills);
  }

  /**
   * Get daily breakdown
   */
  getDailyBreakdown(): Map<number, number> {
    return new Map(this.dailyNoFills);
  }

  /**
   * Get no-fills by source
   */
  getNoFillsBySource(): Map<string, number> {
    return new Map(this.noFillsBySource);
  }

  /**
   * Get no-fills by placement
   */
  getNoFillsByPlacement(): Map<string, number> {
    return new Map(this.noFillsByPlacement);
  }

  /**
   * Get no-fills by reason
   */
  getNoFillsByReason(): Map<NoFillReason, number> {
    return new Map(this.noFillsByReason);
  }

  /**
   * Get events in a time range
   */
  getEvents(from: number, to: number): NoFillEvent[] {
    return this.events.filter((e) => e.timestamp >= from && e.timestamp <= to);
  }

  /**
   * Get recent events
   */
  getRecentEvents(count: number = 100): NoFillEvent[] {
    return this.events.slice(-count);
  }

  /**
   * Get detected patterns
   */
  getDetectedPatterns(): NoFillPattern[] {
    return [...this.detectedPatterns];
  }

  /**
   * Add pattern listener
   */
  addPatternListener(listener: NoFillPatternListener): void {
    this.patternListeners.push(listener);
  }

  /**
   * Remove pattern listener
   */
  removePatternListener(listener: NoFillPatternListener): void {
    const index = this.patternListeners.indexOf(listener);
    if (index !== -1) {
      this.patternListeners.splice(index, 1);
    }
  }

  /**
   * Clear all data
   */
  clear(): void {
    this.events = [];
    this.totalNoFills = 0;
    this.totalLatencyMs = 0;
    this.hourlyNoFills.clear();
    this.dailyNoFills.clear();
    this.noFillsBySource.clear();
    this.noFillsByPlacement.clear();
    this.noFillsByReason.clear();
    this.consecutiveNoFillsBySource.clear();
    this.detectedPatterns = [];
  }

  /**
   * Update configuration
   */
  updateConfiguration(config: Partial<NoFillTrackerConfig>): void {
    this.config = { ...this.config, ...config };
  }

  /**
   * Get current configuration
   */
  getConfiguration(): NoFillTrackerConfig {
    return { ...this.config };
  }

  private cleanupIfNeeded(): void {
    // Remove events exceeding retention limit
    const cutoff = Date.now() - this.config.maxRetentionHours * 3600000;
    this.events = this.events.filter((e) => e.timestamp >= cutoff);

    // Trim to max events
    if (this.events.length > this.config.maxEventsRetained) {
      const excess = this.events.length - this.config.maxEventsRetained;
      this.events.splice(0, excess);
    }
  }

  private detectPatternsForEvent(event: NoFillEvent): void {
    // Check for consecutive failures
    const consecutive =
      this.consecutiveNoFillsBySource.get(event.sourceId) || 0;
    if (consecutive >= this.config.consecutiveFailureThreshold) {
      const severity =
        consecutive >= 10
          ? PatternSeverity.HIGH
          : consecutive >= 5
            ? PatternSeverity.MEDIUM
            : PatternSeverity.LOW;

      const pattern: NoFillPattern = {
        type: PatternType.CONSECUTIVE_FAILURES,
        severity,
        description: `Source ${event.sourceId} has ${consecutive} consecutive no-fills`,
        detectedAt: Date.now(),
        affectedSourceId: event.sourceId,
      };
      this.addPattern(pattern);
    }

    // Check for elevated rate on this source
    const sourceTotal = this.noFillsBySource.get(event.sourceId) || 0;
    const totalAttempts = Math.max(1, this.totalNoFills);
    const sourceRate = sourceTotal / totalAttempts;

    if (sourceRate > this.config.elevatedRateThreshold && sourceTotal > 10) {
      const severity =
        sourceRate > 0.8 ? PatternSeverity.CRITICAL : PatternSeverity.MEDIUM;

      const pattern: NoFillPattern = {
        type: PatternType.SOURCE_SPECIFIC,
        severity,
        description: `Source ${event.sourceId} has ${Math.round(sourceRate * 100)}% no-fill rate`,
        detectedAt: Date.now(),
        affectedSourceId: event.sourceId,
      };
      this.addPattern(pattern);
    }

    // Check for placement-specific issues
    const placementTotal = this.noFillsByPlacement.get(event.placementId) || 0;
    if (placementTotal > 20) {
      const placementRate = placementTotal / Math.max(1, this.totalNoFills);
      if (placementRate > 0.3) {
        const pattern: NoFillPattern = {
          type: PatternType.PLACEMENT_SPECIFIC,
          severity: PatternSeverity.MEDIUM,
          description: `Placement ${event.placementId} accounts for ${Math.round(placementRate * 100)}% of no-fills`,
          detectedAt: Date.now(),
          affectedPlacementId: event.placementId,
        };
        this.addPattern(pattern);
      }
    }
  }

  private addPattern(pattern: NoFillPattern): void {
    // Avoid duplicate patterns within 5 minutes
    const recentCutoff = Date.now() - 300000;
    const isDuplicate = this.detectedPatterns.some(
      (existing) =>
        existing.type === pattern.type &&
        existing.affectedSourceId === pattern.affectedSourceId &&
        existing.affectedPlacementId === pattern.affectedPlacementId &&
        existing.detectedAt > recentCutoff
    );

    if (!isDuplicate) {
      this.detectedPatterns.push(pattern);

      // Trim old patterns
      const oneDayAgo = Date.now() - 86400000;
      this.detectedPatterns = this.detectedPatterns.filter(
        (p) => p.detectedAt >= oneDayAgo
      );

      // Notify listeners
      for (const listener of this.patternListeners) {
        try {
          listener(pattern);
        } catch {
          // Ignore listener errors
        }
      }
    }
  }
}

/**
 * Analyzer for no-fill patterns and recommendations
 */
export class NoFillAnalyzer {
  private tracker: NoFillTracker;

  constructor(tracker?: NoFillTracker) {
    this.tracker = tracker || NoFillTracker.getInstance();
  }

  /**
   * Analyze current no-fill patterns
   */
  analyze(): {
    summary: string;
    recommendations: string[];
    healthScore: number;
    problematicSources: string[];
    problematicPlacements: string[];
  } {
    const stats = this.tracker.getStats();
    const patterns = this.tracker.getDetectedPatterns();

    const recommendations: string[] = [];
    const problematicSources: string[] = [];
    const problematicPlacements: string[] = [];

    // Analyze top sources
    for (const [source, count] of stats.topSources) {
      const percentage = (count / Math.max(1, stats.totalNoFills)) * 100;
      if (percentage > 30) {
        problematicSources.push(source);
        recommendations.push(
          `Consider deprioritizing source '${source}' (accounts for ${Math.round(percentage)}% of no-fills)`
        );
      }
    }

    // Analyze top placements
    for (const [placement, count] of stats.topPlacements) {
      const percentage = (count / Math.max(1, stats.totalNoFills)) * 100;
      if (percentage > 30) {
        problematicPlacements.push(placement);
        recommendations.push(
          `Review placement '${placement}' configuration (accounts for ${Math.round(percentage)}% of no-fills)`
        );
      }
    }

    // Analyze reasons
    let topReason: [NoFillReason, number] | undefined;
    let maxCount = 0;
    for (const [reason, count] of stats.topReasons) {
      if (count > maxCount) {
        maxCount = count;
        topReason = [reason, count];
      }
    }

    if (topReason) {
      const percentage =
        (topReason[1] / Math.max(1, stats.totalNoFills)) * 100;
      if (percentage > 40) {
        switch (topReason[0]) {
          case NoFillReason.TIMEOUT:
            recommendations.push(
              'Consider increasing timeout thresholds or improving network conditions'
            );
            break;
          case NoFillReason.NO_INVENTORY:
            recommendations.push(
              'Explore additional demand sources to improve fill rate'
            );
            break;
          case NoFillReason.NETWORK_ERROR:
            recommendations.push(
              'Review network connectivity and retry strategies'
            );
            break;
          case NoFillReason.FREQUENCY_CAP:
            recommendations.push(
              'Review frequency cap settings - may be too restrictive'
            );
            break;
          default:
            recommendations.push(
              `Investigate high occurrence of '${topReason[0]}' errors`
            );
        }
      }
    }

    // Calculate health score
    let healthScore = 100;

    // Deduct for high no-fill rate
    if (stats.noFillRate > 10) {
      healthScore -= Math.min(50, stats.noFillRate * 2);
    }

    // Deduct for detected patterns
    const criticalPatterns = patterns.filter(
      (p) => p.severity === PatternSeverity.CRITICAL
    ).length;
    const highPatterns = patterns.filter(
      (p) => p.severity === PatternSeverity.HIGH
    ).length;
    healthScore -= criticalPatterns * 15;
    healthScore -= highPatterns * 10;
    healthScore = Math.max(0, Math.min(100, healthScore));

    // Generate summary
    let summary: string;
    if (healthScore >= 80) {
      summary = `No-fill tracking is healthy with ${stats.totalNoFills} events recorded.`;
    } else if (healthScore >= 50) {
      summary = `Moderate no-fill issues detected. ${recommendations.length} recommendations available.`;
    } else {
      summary =
        'Critical no-fill issues detected! Immediate attention recommended.';
    }

    return {
      summary,
      recommendations,
      healthScore,
      problematicSources,
      problematicPlacements,
    };
  }
}

File: sdk/core/ios/Sources/Mediation/NoFillTracker.swift
import Foundation

// MARK: - NoFillReason

/// Reasons for ad no-fill
public enum NoFillReason: String, Codable {
    case timeout = "timeout"
    case noInventory = "no_inventory"
    case networkError = "network_error"
    case policyViolation = "policy_violation"
    case frequencyCap = "frequency_cap"
    case geographicRestriction = "geographic_restriction"
    case budgetExhausted = "budget_exhausted"
    case malformedResponse = "malformed_response"
    case serverError = "server_error"
    case unknown = "unknown"
}

// MARK: - NoFillEvent

/// A single no-fill event
public struct NoFillEvent: Codable {
    public let sourceId: String
    public let placementId: String
    public let reason: NoFillReason
    public let timestamp: Date
    public let latencyMs: Int64
    public let metadata: [String: String]
    
    public init(
        sourceId: String,
        placementId: String,
        reason: NoFillReason,
        timestamp: Date = Date(),
        latencyMs: Int64 = 0,
        metadata: [String: String] = [:]
    ) {
        self.sourceId = sourceId
        self.placementId = placementId
        self.reason = reason
        self.timestamp = timestamp
        self.latencyMs = latencyMs
        self.metadata = metadata
    }
}

// MARK: - NoFillStats

/// Statistics for no-fill events
public struct NoFillStats {
    public let totalNoFills: Int
    public let noFillRate: Double // Per minute
    public let averageLatencyMs: Double
    public let topReasons: [NoFillReason: Int]
    public let topSources: [String: Int]
    public let topPlacements: [String: Int]
    public let hourlyBreakdown: [Int: Int]
    public let dailyBreakdown: [Int: Int]
}

// MARK: - NoFillPattern

/// Detected no-fill pattern
public struct NoFillPattern {
    public let type: PatternType
    public let severity: Severity
    public let description: String
    public let detectedAt: Date
    public let affectedSourceId: String?
    public let affectedPlacementId: String?
    
    public enum PatternType: String {
        case elevatedRate = "elevated_rate"
        case sourceSpecific = "source_specific"
        case placementSpecific = "placement_specific"
        case timeBasedSpike = "time_based_spike"
        case consecutiveFailures = "consecutive_failures"
    }
    
    public enum Severity: String {
        case low = "low"
        case medium = "medium"
        case high = "high"
        case critical = "critical"
    }
}

// MARK: - NoFillTracker

/// Tracks no-fill events for analytics and pattern detection
public final class NoFillTracker {
    
    // MARK: - Singleton
    
    public static let shared = NoFillTracker()
    
    // MARK: - Configuration
    
    public struct Configuration {
        public var maxEventsRetained: Int = 10_000
        public var maxRetentionHours: Int = 24
        public var elevatedRateThreshold: Double = 0.5 // 50% no-fill rate
        public var patternDetectionEnabled: Bool = true
        public var consecutiveFailureThreshold: Int = 5
        
        public static let `default` = Configuration()
    }
    
    // MARK: - Properties
    
    private var config: Configuration
    private var events: [NoFillEvent] = []
    private let lock = NSLock()
    
    // Aggregated counters
    private var totalNoFills: Int = 0
    private var totalLatencyMs: Int64 = 0
    
    // Hourly and daily breakdowns
    private var hourlyNoFills: [Int: Int] = [:] // Hour of day (0-23) -> count
    private var dailyNoFills: [Int: Int] = [:] // Day of week (1-7) -> count
    
    // Per-source and per-placement counters
    private var noFillsBySource: [String: Int] = [:]
    private var noFillsByPlacement: [String: Int] = [:]
    private var noFillsByReason: [NoFillReason: Int] = [:]
    
    // Pattern detection
    private var consecutiveNoFillsBySource: [String: Int] = [:]
    private var detectedPatterns: [NoFillPattern] = []
    
    // Listeners
    private var patternListeners: [(NoFillPattern) -> Void] = []
    
    // MARK: - Initialization
    
    private init(config: Configuration = .default) {
        self.config = config
    }
    
    // MARK: - Public API
    
    /// Record a no-fill event
    public func recordNoFill(
        sourceId: String,
        placementId: String,
        reason: NoFillReason,
        latencyMs: Int64 = 0,
        metadata: [String: String] = [:]
    ) {
        let event = NoFillEvent(
            sourceId: sourceId,
            placementId: placementId,
            reason: reason,
            latencyMs: latencyMs,
            metadata: metadata
        )
        
        lock.lock()
        defer { lock.unlock() }
        
        // Add event
        events.append(event)
        
        // Update counters
        totalNoFills += 1
        totalLatencyMs += latencyMs
        
        // Update source counter
        noFillsBySource[sourceId, default: 0] += 1
        
        // Update placement counter
        noFillsByPlacement[placementId, default: 0] += 1
        
        // Update reason counter
        noFillsByReason[reason, default: 0] += 1
        
        // Update hourly breakdown
        let calendar = Calendar.current
        let hour = calendar.component(.hour, from: event.timestamp)
        hourlyNoFills[hour, default: 0] += 1
        
        // Update daily breakdown
        let weekday = calendar.component(.weekday, from: event.timestamp)
        dailyNoFills[weekday, default: 0] += 1
        
        // Update consecutive failures
        consecutiveNoFillsBySource[sourceId, default: 0] += 1
        
        // Cleanup old events if needed
        cleanupIfNeeded()
        
        // Detect patterns
        if config.patternDetectionEnabled {
            detectPatternsForEvent(event)
        }
    }
    
    /// Record a successful fill (resets consecutive failure counter)
    public func recordFill(sourceId: String) {
        lock.lock()
        defer { lock.unlock() }
        
        consecutiveNoFillsBySource[sourceId] = 0
    }
    
    /// Get current statistics
    public func getStats() -> NoFillStats {
        lock.lock()
        defer { lock.unlock() }
        
        let avgLatency = totalNoFills > 0 
            ? Double(totalLatencyMs) / Double(totalNoFills) 
            : 0.0
        
        // Calculate rate per minute (based on last hour)
        let oneHourAgo = Date().addingTimeInterval(-3600)
        let recentEvents = events.filter { $0.timestamp >= oneHourAgo }
        let ratePerMinute = Double(recentEvents.count) / 60.0
        
        // Get top reasons
        let topReasons = Dictionary(
            noFillsByReason.sorted { $0.value > $1.value }.prefix(5),
            uniquingKeysWith: { first, _ in first }
        )
        
        // Get top sources
        let topSources = Dictionary(
            noFillsBySource.sorted { $0.value > $1.value }.prefix(5),
            uniquingKeysWith: { first, _ in first }
        )
        
        // Get top placements
        let topPlacements = Dictionary(
            noFillsByPlacement.sorted { $0.value > $1.value }.prefix(5),
            uniquingKeysWith: { first, _ in first }
        )
        
        return NoFillStats(
            totalNoFills: totalNoFills,
            noFillRate: ratePerMinute,
            averageLatencyMs: avgLatency,
            topReasons: topReasons,
            topSources: topSources,
            topPlacements: topPlacements,
            hourlyBreakdown: hourlyNoFills,
            dailyBreakdown: dailyNoFills
        )
    }
    
    /// Get hourly breakdown
    public func getHourlyBreakdown() -> [Int: Int] {
        lock.lock()
        defer { lock.unlock() }
        return hourlyNoFills
    }
    
    /// Get daily breakdown
    public func getDailyBreakdown() -> [Int: Int] {
        lock.lock()
        defer { lock.unlock() }
        return dailyNoFills
    }
    
    /// Get no-fills by source
    public func getNoFillsBySource() -> [String: Int] {
        lock.lock()
        defer { lock.unlock() }
        return noFillsBySource
    }
    
    /// Get no-fills by placement
    public func getNoFillsByPlacement() -> [String: Int] {
        lock.lock()
        defer { lock.unlock() }
        return noFillsByPlacement
    }
    
    /// Get no-fills by reason
    public func getNoFillsByReason() -> [NoFillReason: Int] {
        lock.lock()
        defer { lock.unlock() }
        return noFillsByReason
    }
    
    /// Get events for a specific time range
    public func getEvents(from start: Date, to end: Date) -> [NoFillEvent] {
        lock.lock()
        defer { lock.unlock() }
        
        return events.filter { $0.timestamp >= start && $0.timestamp <= end }
    }
    
    /// Get recent events
    public func getRecentEvents(count: Int = 100) -> [NoFillEvent] {
        lock.lock()
        defer { lock.unlock() }
        
        return Array(events.suffix(count))
    }
    
    /// Get detected patterns
    public func getDetectedPatterns() -> [NoFillPattern] {
        lock.lock()
        defer { lock.unlock() }
        return detectedPatterns
    }
    
    /// Add pattern listener
    public func addPatternListener(_ listener: @escaping (NoFillPattern) -> Void) {
        lock.lock()
        defer { lock.unlock() }
        patternListeners.append(listener)
    }
    
    /// Clear all data
    public func clear() {
        lock.lock()
        defer { lock.unlock() }
        
        events.removeAll()
        totalNoFills = 0
        totalLatencyMs = 0
        hourlyNoFills.removeAll()
        dailyNoFills.removeAll()
        noFillsBySource.removeAll()
        noFillsByPlacement.removeAll()
        noFillsByReason.removeAll()
        consecutiveNoFillsBySource.removeAll()
        detectedPatterns.removeAll()
    }
    
    /// Update configuration
    public func updateConfiguration(_ config: Configuration) {
        lock.lock()
        defer { lock.unlock() }
        self.config = config
    }
    
    // MARK: - Private Methods
    
    private func cleanupIfNeeded() {
        // Remove events exceeding retention limit
        let cutoff = Date().addingTimeInterval(-Double(config.maxRetentionHours) * 3600)
        events.removeAll { $0.timestamp < cutoff }
        
        // Trim to max events
        if events.count > config.maxEventsRetained {
            let excess = events.count - config.maxEventsRetained
            events.removeFirst(excess)
        }
    }
    
    private func detectPatternsForEvent(_ event: NoFillEvent) {
        // Check for consecutive failures
        if let consecutive = consecutiveNoFillsBySource[event.sourceId],
           consecutive >= config.consecutiveFailureThreshold {
            let pattern = NoFillPattern(
                type: .consecutiveFailures,
                severity: consecutive >= 10 ? .high : .medium,
                description: "Source \(event.sourceId) has \(consecutive) consecutive no-fills",
                detectedAt: Date(),
                affectedSourceId: event.sourceId,
                affectedPlacementId: nil
            )
            addPattern(pattern)
        }
        
        // Check for elevated rate on this source
        let sourceTotal = noFillsBySource[event.sourceId] ?? 0
        let totalAttempts = max(1, totalNoFills) // Simplified - in real impl would track fills too
        let sourceRate = Double(sourceTotal) / Double(totalAttempts)
        
        if sourceRate > config.elevatedRateThreshold && sourceTotal > 10 {
            let pattern = NoFillPattern(
                type: .sourceSpecific,
                severity: sourceRate > 0.8 ? .critical : .medium,
                description: "Source \(event.sourceId) has \(Int(sourceRate * 100))% no-fill rate",
                detectedAt: Date(),
                affectedSourceId: event.sourceId,
                affectedPlacementId: nil
            )
            addPattern(pattern)
        }
        
        // Check for placement-specific issues
        let placementTotal = noFillsByPlacement[event.placementId] ?? 0
        if placementTotal > 20 {
            let placementRate = Double(placementTotal) / Double(max(1, totalNoFills))
            if placementRate > 0.3 {
                let pattern = NoFillPattern(
                    type: .placementSpecific,
                    severity: .medium,
                    description: "Placement \(event.placementId) accounts for \(Int(placementRate * 100))% of no-fills",
                    detectedAt: Date(),
                    affectedSourceId: nil,
                    affectedPlacementId: event.placementId
                )
                addPattern(pattern)
            }
        }
    }
    
    private func addPattern(_ pattern: NoFillPattern) {
        // Avoid duplicate patterns within 5 minutes
        let recentCutoff = Date().addingTimeInterval(-300)
        let isDuplicate = detectedPatterns.contains { existing in
            existing.type == pattern.type &&
            existing.affectedSourceId == pattern.affectedSourceId &&
            existing.affectedPlacementId == pattern.affectedPlacementId &&
            existing.detectedAt > recentCutoff
        }
        
        if !isDuplicate {
            detectedPatterns.append(pattern)
            
            // Trim old patterns
            let oneDayAgo = Date().addingTimeInterval(-86400)
            detectedPatterns.removeAll { $0.detectedAt < oneDayAgo }
            
            // Notify listeners
            for listener in patternListeners {
                listener(pattern)
            }
        }
    }
}

// MARK: - NoFillAnalyzer

/// Analyzes no-fill patterns and provides recommendations
public final class NoFillAnalyzer {
    
    public struct Analysis {
        public let summary: String
        public let recommendations: [String]
        public let healthScore: Double // 0-100
        public let problematicSources: [String]
        public let problematicPlacements: [String]
    }
    
    private let tracker: NoFillTracker
    
    public init(tracker: NoFillTracker = .shared) {
        self.tracker = tracker
    }
    
    /// Analyze current no-fill patterns
    public func analyze() -> Analysis {
        let stats = tracker.getStats()
        let patterns = tracker.getDetectedPatterns()
        
        var recommendations: [String] = []
        var problematicSources: [String] = []
        var problematicPlacements: [String] = []
        
        // Analyze top sources
        for (source, count) in stats.topSources {
            let percentage = Double(count) / Double(max(1, stats.totalNoFills)) * 100
            if percentage > 30 {
                problematicSources.append(source)
                recommendations.append("Consider deprioritizing source '\(source)' (accounts for \(Int(percentage))% of no-fills)")
            }
        }
        
        // Analyze top placements
        for (placement, count) in stats.topPlacements {
            let percentage = Double(count) / Double(max(1, stats.totalNoFills)) * 100
            if percentage > 30 {
                problematicPlacements.append(placement)
                recommendations.append("Review placement '\(placement)' configuration (accounts for \(Int(percentage))% of no-fills)")
            }
        }
        
        // Analyze reasons
        if let (topReason, count) = stats.topReasons.max(by: { $0.value < $1.value }) {
            let percentage = Double(count) / Double(max(1, stats.totalNoFills)) * 100
            if percentage > 40 {
                switch topReason {
                case .timeout:
                    recommendations.append("Consider increasing timeout thresholds or improving network conditions")
                case .noInventory:
                    recommendations.append("Explore additional demand sources to improve fill rate")
                case .networkError:
                    recommendations.append("Review network connectivity and retry strategies")
                case .frequencyCap:
                    recommendations.append("Review frequency cap settings - may be too restrictive")
                default:
                    recommendations.append("Investigate high occurrence of '\(topReason.rawValue)' errors")
                }
            }
        }
        
        // Calculate health score (100 = perfect, 0 = critical issues)
        var healthScore: Double = 100.0
        
        // Deduct for high no-fill rate
        let noFillRate = stats.noFillRate
        if noFillRate > 10 {
            healthScore -= min(50, noFillRate * 2)
        }
        
        // Deduct for detected patterns
        let criticalPatterns = patterns.filter { $0.severity == .critical }.count
        let highPatterns = patterns.filter { $0.severity == .high }.count
        healthScore -= Double(criticalPatterns * 15)
        healthScore -= Double(highPatterns * 10)
        healthScore = max(0, min(100, healthScore))
        
        // Generate summary
        let summary: String
        if healthScore >= 80 {
            summary = "No-fill tracking is healthy with \(stats.totalNoFills) events recorded."
        } else if healthScore >= 50 {
            summary = "Moderate no-fill issues detected. \(recommendations.count) recommendations available."
        } else {
            summary = "Critical no-fill issues detected! Immediate attention recommended."
        }
        
        return Analysis(
            summary: summary,
            recommendations: recommendations,
            healthScore: healthScore,
            problematicSources: problematicSources,
            problematicPlacements: problematicPlacements
        )
    }
}
